File: apps/backend-node/src/features/properties/property.service.ts
import { Injectable } from '@nestjs/common';
import type { properties } from '@prisma/client';
import { Prisma } from '@prisma/client';

import { PrismaService } from '../../prisma/prisma.service';
import type { AuthUser } from '../../shared/auth/auth-user.model';
import { AuthorizationService } from '../../shared/auth/authorization.service';
import {
  BadRequestError,
  ForbiddenError,
  NotFoundError,
} from '../../shared/errors';
import { StorageUrlResolver } from '../../shared/storage/storage-url.resolver';
import type { CreateRoomRequest } from './dto/create-room-request';
import type { CreatedId } from './dto/created-id';
import type { HostPropertyDetail } from './dto/host-property-detail';
import type { HostPropertyRequest } from './dto/host-property-request';
import type { HostRoomDto } from './dto/host-room';
import type { PropertyAmenityDto } from './dto/property-amenity';
import type { PropertyDetail } from './dto/property-detail';
import type { PropertyStatusChange } from './dto/property-status-change';
import type { SetPropertyLocationRequest } from './dto/set-property-location-request';
import {
  createPropertySlug,
  mapAmenityDto,
  mapHostPropertyDetail,
  mapHostRoom,
  mapPropertyDetail,
  normalizeTags,
  propertyStatusToApi,
  requestToPropertyFields,
} from './mappers/property.mapper';
import { PropertyRepository } from './property.repository';

@Injectable()
export class PropertyService {
  constructor(
    private readonly prisma: PrismaService,
    private readonly repository: PropertyRepository,
    private readonly auth: AuthorizationService,
    private readonly storage: StorageUrlResolver,
  ) {}

  async getPublishedProperty(propertyId: string): Promise<PropertyDetail> {
    const propertyRow = await this.repository.getPublishedProperty(propertyId);
    if (!propertyRow || propertyRow.monthly_price_min == null) {
      throw new NotFoundError('Property not found');
    }

    const coordinates =
      await this.repository.getPublishedCoordinates(propertyId);
    const amenities = await this.repository.getPropertyAmenities(propertyId);
    const hostDisplayName = await this.repository.getHostDisplayName(
      propertyRow.host_id,
    );

    return mapPropertyDetail(propertyRow, {
      hostDisplayName,
      latitude: coordinates?.latitude ?? null,
      longitude: coordinates?.longitude ?? null,
      amenities,
      storage: this.storage,
    });
  }

  async createProperty(
    user: AuthUser,
    request: HostPropertyRequest,
  ): Promise<CreatedId> {
    const hostId = await this.requireHostId(user);
    const slug = createPropertySlug(request.title);
    const amenityIds = request.amenityIds ?? [];

    const propertyId = await this.prisma.$transaction(async (tx) => {
      const propertyRow = await tx.properties.create({
        data: {
          host_id: hostId,
          slug,
          status: 'draft',
          country: 'KR',
          tags: normalizeTags(request.tags),
          ...requestToPropertyFields(request),
        },
      });

      await this.applyLocationIfPresentTx(tx, propertyRow.id, request);
      await this.syncAmenitiesTx(tx, propertyRow.id, amenityIds);
      return propertyRow.id;
    });

    return { id: propertyId };
  }

  async updateProperty(
    user: AuthUser,
    propertyId: string,
    request: HostPropertyRequest,
  ): Promise<HostPropertyDetail> {
    await this.requireMutableProperty(user, propertyId);
    const amenityIds = request.amenityIds ?? [];

    await this.prisma.$transaction(async (tx) => {
      const existing = await tx.properties.findFirst({
        where: {
          id: propertyId,
          deleted_at: null,
        },
      });

      if (!existing) {
        throw new NotFoundError('Property not found');
      }

      await tx.properties.update({
        where: { id: propertyId },
        data: {
          ...requestToPropertyFields(request),
          tags: normalizeTags(request.tags),
          updated_at: new Date(),
        },
      });

      await this.applyLocationIfPresentTx(tx, propertyId, request);
      await this.syncAmenitiesTx(tx, propertyId, amenityIds);
    });

    return this.loadHostPropertyDetail(propertyId);
  }

  async setLocation(
    user: AuthUser,
    propertyId: string,
    request: SetPropertyLocationRequest,
  ): Promise<void> {
    await this.requireMutableProperty(user, propertyId);

    const updated = await this.repository.setLocation(
      propertyId,
      request.latitude,
      request.longitude,
    );

    if (!updated) {
      throw new BadRequestError('Property location cannot be updated');
    }
  }

  async submitForReview(
    user: AuthUser,
    propertyId: string,
  ): Promise<PropertyStatusChange> {
    await this.auth.requireHostOfProperty(user, propertyId);

    const propertyRow = await this.repository.getHostProperty(propertyId);
    if (!propertyRow) {
      throw new NotFoundError('Property not found');
    }

    if (propertyRow.status !== 'draft') {
      throw new BadRequestError(
        'Property must be in draft status to submit for review',
      );
    }

    if ((await this.repository.countAvailableRooms(propertyId)) === 0) {
      throw new BadRequestError(
        'Add at least one available room before submitting',
      );
    }

    if (!(await this.repository.hasLocation(propertyId))) {
      throw new BadRequestError(
        'Geocode the property address before submitting',
      );
    }

    const updated = await this.repository.submitForReview(propertyId);
    if (!updated) {
      throw new BadRequestError(
        'Property must be in draft status to submit for review',
      );
    }

    return {
      id: updated.id,
      status: propertyStatusToApi(updated.status),
    };
  }

  async createRoom(
    user: AuthUser,
    propertyId: string,
    request: CreateRoomRequest,
  ): Promise<HostRoomDto> {
    await this.requireMutableProperty(user, propertyId);
    const room = await this.repository.createRoom(propertyId, request);
    return mapHostRoom(room);
  }

  async deleteRoom(user: AuthUser, roomId: string): Promise<void> {
    const propertyId = await this.repository.findRoomPropertyId(roomId);
    if (!propertyId) {
      throw new NotFoundError('Room not found');
    }

    if (
      !(await this.auth.isHostOfProperty(user.id, propertyId)) &&
      !(await this.auth.isAdmin(user.id))
    ) {
      throw new ForbiddenError(
        'Only the host of this property can perform this action',
      );
    }

    const deleted = await this.repository.deleteRoom(roomId);
    if (!deleted) {
      throw new NotFoundError('Room not found');
    }
  }

  async listAmenities(): Promise<PropertyAmenityDto[]> {
    const amenities = await this.repository.findAllAmenities();
    return amenities.map(mapAmenityDto);
  }

  private async loadHostPropertyDetail(
    propertyId: string,
  ): Promise<HostPropertyDetail> {
    const propertyRow = await this.repository.getHostProperty(propertyId);
    if (!propertyRow) {
      throw new NotFoundError('Property not found');
    }

    const coordinates = await this.repository.getCoordinates(propertyId);
    const amenityIds = await this.repository.getPropertyAmenityIds(propertyId);

    return mapHostPropertyDetail(propertyRow, {
      latitude: coordinates?.latitude ?? null,
      longitude: coordinates?.longitude ?? null,
      amenityIds,
    });
  }

  private async requireHostId(user: AuthUser): Promise<string> {
    const hostId = await this.auth.getHostIdForProfile(user.id);
    if (!hostId) {
      throw new ForbiddenError(
        'Host profile is required before creating listings',
      );
    }
    return hostId;
  }

  private async requireMutableProperty(
    user: AuthUser,
    propertyId: string,
  ): Promise<properties> {
    const propertyRow = await this.repository.getHostProperty(propertyId);
    if (!propertyRow) {
      throw new NotFoundError('Property not found');
    }

    if (await this.auth.isAdmin(user.id)) {
      return propertyRow;
    }

    if (!(await this.auth.isHostOfProperty(user.id, propertyId))) {
      throw new ForbiddenError(
        'Only the host of this property can perform this action',
      );
    }

    if (
      propertyRow.status !== 'draft' &&
      propertyRow.status !== 'pending_review'
    ) {
      throw new BadRequestError(
        'Property can only be updated while draft or pending review',
      );
    }

    return propertyRow;
  }

  private async applyLocationIfPresentTx(
    tx: Prisma.TransactionClient,
    propertyId: string,
    request: HostPropertyRequest,
  ): Promise<void> {
    if (request.latitude == null || request.longitude == null) {
      return;
    }

    const rows = await tx.$queryRaw<{ id: string }[]>`
      UPDATE public.properties
      SET location = extensions.st_setsrid(
        extensions.st_makepoint(${request.longitude}, ${request.latitude}),
        4326
      )::extensions.geography
      WHERE id = ${propertyId}::uuid
        AND deleted_at IS NULL
        AND status IN ('draft', 'pending_review')
      RETURNING id
    `;

    if (rows.length === 0) {
      throw new BadRequestError('Property location cannot be updated');
    }
  }

  private async syncAmenitiesTx(
    tx: Prisma.TransactionClient,
    propertyId: string,
    amenityIds: string[],
  ): Promise<void> {
    await tx.property_amenities.deleteMany({
      where: { property_id: propertyId },
    });

    if (amenityIds.length === 0) {
      return;
    }

    await tx.property_amenities.createMany({
      data: amenityIds.map((amenityId) => ({
        property_id: propertyId,
        amenity_id: amenityId,
      })),
    });
  }
}

File: apps/backend-java/src/main/java/com/housingplatform/properties/PropertyService.java
package com.housingplatform.properties;

import com.housingplatform.auth.error.BadRequestException;
import com.housingplatform.auth.error.ForbiddenException;
import com.housingplatform.auth.error.NotFoundException;
import com.housingplatform.auth.model.AuthenticatedUser;
import com.housingplatform.auth.service.AuthorizationService;
import com.housingplatform.persistence.entity.Property;
import com.housingplatform.persistence.repository.PropertyRepository;
import com.housingplatform.persistence.enums.PropertyStatus;
import com.housingplatform.properties.dto.CreateRoomRequest;
import com.housingplatform.properties.dto.CreatedId;
import com.housingplatform.properties.dto.HostPropertyDetail;
import com.housingplatform.properties.dto.HostPropertyRequest;
import com.housingplatform.properties.dto.HostRoomDto;
import com.housingplatform.properties.dto.PropertyAmenityDto;
import com.housingplatform.properties.dto.PropertyDetail;
import com.housingplatform.properties.dto.PropertySearchQuery;
import com.housingplatform.properties.dto.PropertySearchResult;
import com.housingplatform.properties.dto.PropertyStatusChange;
import com.housingplatform.properties.dto.SetPropertyLocationRequest;
import com.housingplatform.properties.mapper.PropertyMapper;
import com.housingplatform.shared.StorageUrlResolver;
import java.util.List;
import java.util.UUID;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class PropertyService {

  private final PropertyRepository propertyRepository;
  private final AuthorizationService authorizationService;
  private final StorageUrlResolver storageUrlResolver;

  public PropertyService(
      PropertyRepository propertyRepository,
      AuthorizationService authorizationService,
      StorageUrlResolver storageUrlResolver) {
    this.propertyRepository = propertyRepository;
    this.authorizationService = authorizationService;
    this.storageUrlResolver = storageUrlResolver;
  }

  @Transactional(readOnly = true)
  public PropertySearchResult search(PropertySearchQuery query) {
    validateSearchQuery(query);
    var rows =
        propertyRepository.search(
            propertyRepository.buildSearchFilters(query), query.limit(), query.offset());
    int totalCount = rows.isEmpty() ? 0 : rows.getFirst().totalCount();
    List<com.housingplatform.properties.dto.SearchPropertyCard> items =
        rows.stream()
            .map(row -> PropertyMapper.toSearchCard(row, storageUrlResolver))
            .toList();
    return new PropertySearchResult(items, totalCount);
  }

  @Transactional(readOnly = true)
  public PropertyDetail getPublishedProperty(UUID propertyId) {
    Property property =
        propertyRepository
            .findPublishedProperty(propertyId)
            .orElseThrow(() -> new NotFoundException("Property not found"));

    if (property.getMonthlyPriceMin() == null) {
      throw new NotFoundException("Property not found");
    }

    var coordinates = propertyRepository.findPublishedCoordinates(propertyId).orElse(null);
    var amenities = propertyRepository.findPropertyAmenities(propertyId);
    var images = propertyRepository.findPropertyImages(propertyId);
    var rooms = propertyRepository.findPropertyRooms(propertyId);
    String hostDisplayName = propertyRepository.findHostDisplayName(property.getHostId());

    return PropertyMapper.toPropertyDetail(
        property,
        hostDisplayName,
        coordinates == null ? null : coordinates.latitude(),
        coordinates == null ? null : coordinates.longitude(),
        images,
        rooms,
        amenities,
        storageUrlResolver);
  }

  @Transactional
  public CreatedId createProperty(AuthenticatedUser user, HostPropertyRequest request) {
    UUID hostId = requireHostId(user);
    String slug = PropertyMapper.createPropertySlug(request.title());
    List<String> tags = PropertyMapper.normalizeTags(request.tags());
    List<UUID> amenityIds = request.amenityIds() == null ? List.of() : request.amenityIds();

    UUID propertyId = propertyRepository.createProperty(hostId, slug, request, tags);
    applyLocationIfPresent(propertyId, request);
    propertyRepository.syncAmenities(propertyId, amenityIds);
    return new CreatedId(propertyId);
  }

  @Transactional
  public HostPropertyDetail updateProperty(
      AuthenticatedUser user, UUID propertyId, HostPropertyRequest request) {
    requireMutableProperty(user, propertyId);
    List<String> tags = PropertyMapper.normalizeTags(request.tags());
    List<UUID> amenityIds = request.amenityIds() == null ? List.of() : request.amenityIds();

    Property updated = propertyRepository.updateProperty(propertyId, request, tags);
    if (updated == null) {
      throw new NotFoundException("Property not found");
    }

    applyLocationIfPresent(propertyId, request);
    propertyRepository.syncAmenities(propertyId, amenityIds);
    return loadHostPropertyDetail(propertyId);
  }

  @Transactional
  public void setLocation(
      AuthenticatedUser user, UUID propertyId, SetPropertyLocationRequest request) {
    requireMutableProperty(user, propertyId);
    if (!propertyRepository.setLocation(propertyId, request.latitude(), request.longitude())) {
      throw new BadRequestException("Property location cannot be updated");
    }
  }

  @Transactional
  public PropertyStatusChange submitForReview(AuthenticatedUser user, UUID propertyId) {
    authorizationService.requireHostOfProperty(user, propertyId);
    Property property =
        propertyRepository
            .findHostProperty(propertyId)
            .orElseThrow(() -> new NotFoundException("Property not found"));

    if (property.getStatus() != PropertyStatus.draft) {
      throw new BadRequestException("Property must be in draft status to submit for review");
    }
    if (propertyRepository.countAvailableRooms(propertyId) == 0) {
      throw new BadRequestException("Add at least one available room before submitting");
    }
    if (!propertyRepository.hasLocation(propertyId)) {
      throw new BadRequestException("Geocode the property address before submitting");
    }

    Property updated = propertyRepository.submitForReview(propertyId);
    if (updated == null) {
      throw new BadRequestException("Property must be in draft status to submit for review");
    }
    return new PropertyStatusChange(updated.getId(), updated.getStatus());
  }

  @Transactional
  public HostRoomDto createRoom(AuthenticatedUser user, UUID propertyId, CreateRoomRequest request) {
    requireMutableProperty(user, propertyId);
    var row = propertyRepository.createRoom(propertyId, request);
    return PropertyMapper.toHostRoomFromRow(
        row.id(),
        row.name(),
        row.roomType(),
        row.sizeSqm(),
        row.maxOccupancy(),
        row.monthlyPriceKrw(),
        row.status(),
        row.availableFrom());
  }

  @Transactional
  public void deleteRoom(AuthenticatedUser user, UUID roomId) {
    UUID propertyId =
        propertyRepository
            .findRoomPropertyId(roomId)
            .orElseThrow(() -> new NotFoundException("Room not found"));

    if (!authorizationService.isHostOfProperty(user.id(), propertyId)
        && !authorizationService.isAdmin(user.id())) {
      throw new ForbiddenException("Only the host of this property can perform this action");
    }

    if (!propertyRepository.deleteRoom(roomId)) {
      throw new NotFoundException("Room not found");
    }
  }

  @Transactional(readOnly = true)
  public List<PropertyAmenityDto> listAmenities() {
    return propertyRepository.findAllAmenities().stream()
        .map(PropertyMapper::toAmenityDto)
        .toList();
  }

  private HostPropertyDetail loadHostPropertyDetail(UUID propertyId) {
    Property property =
        propertyRepository
            .findHostProperty(propertyId)
            .orElseThrow(() -> new NotFoundException("Property not found"));
    var coordinates = propertyRepository.findCoordinates(propertyId).orElse(null);
    var amenityIds = propertyRepository.findPropertyAmenityIds(propertyId);
    var rooms = propertyRepository.findPropertyRooms(propertyId);
    return PropertyMapper.toHostPropertyDetail(
        property,
        coordinates == null ? null : coordinates.latitude(),
        coordinates == null ? null : coordinates.longitude(),
        amenityIds,
        rooms);
  }

  private UUID requireHostId(AuthenticatedUser user) {
    UUID hostId = authorizationService.getHostIdForProfile(user.id());
    if (hostId == null) {
      throw new ForbiddenException("Host profile is required before creating listings");
    }
    return hostId;
  }

  private Property requireMutableProperty(AuthenticatedUser user, UUID propertyId) {
    Property property =
        propertyRepository
            .findHostProperty(propertyId)
            .orElseThrow(() -> new NotFoundException("Property not found"));

    if (authorizationService.isAdmin(user.id())) {
      return property;
    }

    if (!authorizationService.isHostOfProperty(user.id(), propertyId)) {
      throw new ForbiddenException("Only the host of this property can perform this action");
    }

    if (property.getStatus() != PropertyStatus.draft
        && property.getStatus() != PropertyStatus.pending_review) {
      throw new BadRequestException("Property can only be updated while draft or pending review");
    }

    return property;
  }

  private void applyLocationIfPresent(UUID propertyId, HostPropertyRequest request) {
    if (request.latitude() == null || request.longitude() == null) {
      return;
    }
    if (!propertyRepository.setLocation(propertyId, request.latitude(), request.longitude())) {
      throw new BadRequestException("Property location cannot be updated");
    }
  }

  private static void validateSearchQuery(PropertySearchQuery query) {
    if (query.priceMin() != null
        && query.priceMax() != null
        && query.priceMax() < query.priceMin()) {
      throw new BadRequestException("priceMax must be greater than or equal to priceMin");
    }
    if (query.checkIn() != null
        && query.checkOut() != null
        && !query.checkOut().isAfter(query.checkIn())) {
      throw new BadRequestException("checkOut must be after checkIn");
    }
  }
}
